general/okina.hpp

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#ifndef MFEM_OKINA_HPP
+#define MFEM_OKINA_HPP
+
+#ifdef __NVCC__
+#include <cuda.h>
+#include <helper_cuda.h>
+#include <helper_functions.h>
+#else
+#define __host__
+#define __device__
+#define __kernel__
+#define __constant__
+#endif // __NVCC__
+
+#include <cmath>
+#include <cassert>
+#include <iostream>
+#include <cstring>
+
+#include <stdarg.h>
+#include <signal.h>
+#include <setjmp.h>
+
+// *****************************************************************************
+//#include "/home/camier1/home/stk/stk.hpp"
+
+// *****************************************************************************
+#define MFEM_NAMESPACE namespace mfem {
+#define MFEM_NAMESPACE_END }
+
+// *****************************************************************************
+#include "dbg.hpp"
+#include "memmng.hpp"
+#include "config.hpp"
+#include "kernels.hpp"
+
+// *****************************************************************************
+#define LOG2(X) ((unsigned) (8*sizeof(unsigned long long)-__builtin_clzll((X))))
+#define ISQRT(N) static_cast<unsigned>(sqrt(static_cast<float>(N)))
+#define ICBRT(N) static_cast<unsigned>(cbrt(static_cast<float>(N)))
+#define IROOT(D,N) ((D==1)?N:(D==2)?ISQRT(N):(D==3)?ICBRT(N):0)
+
+// *****************************************************************************
+#define GET_CUDA const bool cuda = mfem::config::Get().Cuda();
+#define GET_ADRS(v) double *d_##v = (double*) mfem::mm::Get().Adrs(v)
```

Comment:
I really don't like doing this as a macro and hiding the fact that you are creating d_ version of the variable.  I would much rather do this as an inlined method that passes back the pointer and lets us select the name.

Reply:
Sure, I don't like it either at all. I'm pushing the preprocessing tool that will add it.